inorbit_edge/missions.py

```diff
@@ -0,0 +1,537 @@
+            ).start()
+
+    def _run_mission_thread(self, mission):
+        mission.execute()
+        self.mission = None
+        self.is_idle.set()
+
+    def wait_until_idle(self, timeout=None):
+        """
+        Waits until the executor is idle.
+        This method is mostly a helper for tests to wait for mission completion.
+        """
+        self.is_idle.wait(timeout)
+        return self.is_idle.is_set()
+
+    def cancel_mission(self, mission_id):
+        with self.mutex:
+            if self.mission is None:
+                self.logger.warning("Can't cancel mission when no mission is running")
+            elif self.mission.id != mission_id and mission_id != "*":
+                self.logger.warning(
+                    f"Can't cancel mission {mission_id} because the id does not match\
+                        running mission {self.mission.id}"
+                )
+            self.mission.cancel()
+            self.mission = None
+            self.is_idle.set()
+
+    def handle_event(self, event):
+        with self.mutex:
+            if self.mission is not None:
+                self.mission.handle_event(event)
+
+
+class Mission:
+    """
+    Provides execution and tracking of a mission
+    """
+
+    def __init__(self, id, program, robot_session):
+        """
+        Initializes the mission from a mission program
+        """
+        self.id = id
+        self.label = program["label"]
+        self.start_ts = int(time.time())
+        self.end_ts = None
+        self.robot_session = robot_session
+        self.defaultStepTimeoutMs = None
+        self.steps = self._build_steps(program)
+        self.state = MISSION_STATE_STARTING
+        self.status = MISSION_STATUS_OK
+        self.current_step_idx = -1
+        self.current_step = None
+        self.data = {}
+        # mutex for state, status and current step and reporting
+        self.mutex = threading.Lock()
+
+    def set_data(self, data: dict):
+        """
+        Adds data to this mission's data
+        """
+        self.data.update(data)
+
+    def execute(self):
+        """
+        Runs this mission's steps
+        """
+        with self.mutex:
+            if self.state == MISSION_STATE_STARTING:
+                self.state = MISSION_STATE_EXECUTING
+        for step_idx in range(0, len(self.steps)):
+            with self.mutex:
+                if self.state != MISSION_STATE_EXECUTING:
+                    # Mission was aborted
+                    break
+                self.current_step_idx = step_idx
+                self.current_step = self.steps[step_idx]
+                self.report()
+            # HACK(mike) sending two reports without waiting can cause issues
+            # in the backend
+            time.sleep(5)
+            try:
+                self.current_step.execute(self)
+                if not self.current_step.success():
+                    raise Exception()
```

Comment:
Raising bare Exceptions is not recommended. I'd change it for a less general `RuntimeError` exception.

I'd also consider defining our own exceptions to improve readability e.g.

```
class FailedMissionStepException(Exception):
    "Raised when mission step fails to execute"
    pass
```